packages/flutter_tools/lib/src/base/logger.dart

```diff
@@ -280,57 +271,138 @@ enum _LogType {
 
+  void _callback(Timer _) {
+    stdout.write('\b${_progress[ticks++ % _progress.length]}');
+  }
+
+  @override
+  void start() {
+    super.start();
+    stdout.write(' ');
+    _callback(null);
     timer = new Timer.periodic(const Duration(milliseconds: 100), _callback);
   }
 
-  static final List<String> _progress = <String>['-', r'\', '|', r'/', '-', r'\', '|', '/'];
+  @override
+  /// Clears the spinner.  After cancel, the cursor will be one space right
+  /// of where it was when [start] was called (assuming no other input).
+  void cancel() {
+    if (timer?.isActive == true) {
+      timer.cancel();
+      // Many terminals do not interpret backspace as deleting a character,
+      // but rather just moving the cursor back one.
+      stdout.write('\b \b');
+    }
+    super.cancel();
+  }
+}
+
+/// Constructor writes [message] to [stdout] with padding, then starts as an
+/// [AnsiSpinner].  On [cancel] or [stop], will call [onFinish].
+/// On [stop], will additionally print out summary information in
+/// milliseconds if [expectSlowOperation] is false, as seconds otherwise.
+class AnsiStatus extends AnsiSpinner {
+  AnsiStatus(this.message, this.expectSlowOperation, this.onFinish, this.padding);
 
   final String message;
   final bool expectSlowOperation;
   final _FinishCallback onFinish;
+  final int padding;
+
   Stopwatch stopwatch;
-  Timer timer;
-  int index = 1;
-  bool live = true;
+  bool _finished = false;
 
-  void _callback(Timer timer) {
-    stdout.write('\b${_progress[index]}');
-    index = ++index % _progress.length;
+  @override
+  /// Writes [message] to [stdout] with padding, then begins spinning.
+  void start() {
+    stopwatch = new Stopwatch()..start();
+    stdout.write('${message.padRight(padding)}     ');
+    assert(!_finished);
+    super.start();
   }
 
   @override
+  /// Calls onFinish.
   void stop() {
-    onFinish();
-
-    if (!live)
-      return;
-    live = false;
+    if (!_finished) {
+      onFinish();
+      _finished = true;
+    }
+    super.stop();
+  }
 
+  @override
+  /// Backs up 4 characters and prints a (minimum) 5 character padded time.  If
+  /// [expectSlowOperation] is true, the time is in seconds; otherwise,
+  /// milliseconds. Only backs up 4 because [AnsiSpinner.cancel] backs up one
+  /// for us.
+  ///
+  /// Example: '\b\b\b\b 0.5s', '\b\b\b\b150ms', '\b\b\b\b1600ms'
+  void summaryInformation() {
     if (expectSlowOperation) {
-      print('\b\b\b\b\b${getElapsedAsSeconds(stopwatch.elapsed).padLeft(5)}');
+      stdout.writeln('\b\b\b\b${getElapsedAsSeconds(stopwatch.elapsed).padLeft(5)}');
     } else {
-      print('\b\b\b\b\b${getElapsedAsMilliseconds(stopwatch.elapsed).padLeft(5)}');
+      stdout.writeln('\b\b\b\b${getElapsedAsMilliseconds(stopwatch.elapsed).padLeft(5)}');
     }
-
-    timer.cancel();
   }
 
   @override
+  /// Calls [onFinish].
   void cancel() {
-    onFinish();
-
-    if (!live)
-      return;
-    live = false;
-
-    print('\b ');
```

Comment:
I think we should probably keep the original behaviour of having a new line after cancel.

Reply:
done